homeassistant/components/time_date/sensor.py

```diff
@@ -46,40 +46,67 @@ async def async_setup_platform(
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up the Time and Date sensor."""
-    if hass.config.time_zone is None:
-        _LOGGER.error("Timezone is not set in Home Assistant configuration")
-        return False
-
-    async_add_entities(
-        [TimeDateSensor(hass, variable) for variable in config[CONF_DISPLAY_OPTIONS]]
+    async_create_issue(
+        hass,
+        HOMEASSISTANT_DOMAIN,
+        f"deprecated_yaml_{DOMAIN}",
+        breaks_in_ha_version="2024.6.0",
+        is_fixable=False,
+        issue_domain=DOMAIN,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+        translation_placeholders={
+            "domain": DOMAIN,
+            "integration_title": "Time & Date",
+        },
     )
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Time & Date sensor."""
+    entities = []
+    for option_type in entry.options[CONF_DISPLAY_OPTIONS]:
+        entities.append(TimeDateSensor(option_type, entry.entry_id))
+    async_add_entities(entities)
 
 
 class TimeDateSensor(SensorEntity):
     """Implementation of a Time and Date sensor."""
 
-    def __init__(self, hass, option_type):
+    _attr_has_entity_name = True
+    _state: str | None = None
+    unsub: Callable[[], None] | None = None
+
+    def __init__(self, option_type: str, entry_id: str) -> None:
         """Initialize the sensor."""
-        self._name = OPTION_TYPES[option_type]
+        self._attr_translation_key = option_type
         self.type = option_type
-        self._state = None
-        self.hass = hass
-        self.unsub = None
+        self.entity_id = ENTITY_ID_FORMAT.format(option_type)
```

Comment:
Why do we overwrite entity id?

Reply:
This allows integrations to suggest an entity id, at the same time I'm not sure why that is added in this PR.

Reply:
It's to avoid having entities like `sensor.time_date_date_time` which makes no sense

Reply:
That's a good point lol

Reply:
We still need to fix the naming issue on this one.
We can't have Names like "Time & Date Date & Time". I'd like to have the device but I'd like to get rid of it's function for the naming but haven't looked too much at it yet.

Reply:
Hmm, there's an architecture discussion for exactly this issue: https://github.com/home-assistant/architecture/discussions/957

Reply:
Yes and no I guess but if we could get placeholders I could at least put `-` in between the device and entity name or such to make it somewhat better looking.
At the moment not having that is a showstopper unless I abuse the entity internals.